File: app/src/main/java/com/hover/stax/channels/ChannelsDropdownAdapter.java
package com.hover.stax.channels;

import static com.hover.stax.utils.Constants.size55;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.core.graphics.drawable.RoundedBitmapDrawable;
import androidx.core.graphics.drawable.RoundedBitmapDrawableFactory;

import com.hover.stax.databinding.StaxSpinnerItemWithLogoBinding;
import com.hover.stax.utils.UIHelper;
import com.squareup.picasso.Picasso;
import com.squareup.picasso.Target;

import java.util.List;

import timber.log.Timber;

public class ChannelsDropdownAdapter extends ArrayAdapter<Channel> {
    private final List<Channel> channels;
    private ViewHolder holder;
    private StaxSpinnerItemWithLogoBinding binding;

    public ChannelsDropdownAdapter(@NonNull List<Channel> channelList, @NonNull Context context) {
        super(context, 0, channelList);
        channels = channelList;
    }

    @NonNull
    @Override
    public View getView(int position, @Nullable View view, @NonNull ViewGroup parent) {
        Channel c = channels.get(position);

        if (view == null) {
            binding = StaxSpinnerItemWithLogoBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
            view = binding.getRoot();
            holder = new ViewHolder(binding);

            view.setTag(holder);
        } else {
            holder = (ViewHolder) view.getTag();
        }

        holder.setChannel(c);

        return view;
    }

    @Override
    public int getCount() {
        return channels.size();
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Nullable
    @Override
    public Channel getItem(int position) {
        return channels.isEmpty() ? null : channels.get(position);
    }

    @Override
    public int getItemViewType(int position) {
        return position;
    }

    public static class ViewHolder implements Target {
        TextView id;
        ImageView logo;
        AppCompatTextView channelText;

        private ViewHolder(StaxSpinnerItemWithLogoBinding withLogoBinding) {
            logo = withLogoBinding.serviceItemImageId;
            channelText = withLogoBinding.serviceItemNameId;
            id = withLogoBinding.serviceItemId;
        }

        @SuppressLint("SetTextI18n")
        private void setChannel(Channel channel) {
            id.setText(Integer.toString(channel.id));
            channelText.setText(channel.toString());
            UIHelper.loadPicasso(channel.logoUrl, size55, this);
        }

        @Override
        public void onBitmapLoaded(Bitmap bitmap, Picasso.LoadedFrom from) {
            RoundedBitmapDrawable d = RoundedBitmapDrawableFactory.create(id.getContext().getResources(), bitmap);
            d.setCircular(true);
            logo.setImageDrawable(d);
        }

        @Override
        public void onBitmapFailed(Exception e, Drawable errorDrawable) {
            Timber.e(e);
        }

        @Override
        public void onPrepareLoad(Drawable placeHolderDrawable) {
        }
    }
}

File: app/src/main/java/com/hover/stax/channels/ChannelsDropdownAdapter.kt
package com.hover.stax.channels

import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import androidx.core.graphics.drawable.RoundedBitmapDrawableFactory
import com.hover.stax.databinding.StaxSpinnerItemWithLogoBinding
import com.hover.stax.utils.Constants
import com.hover.stax.utils.UIHelper
import com.squareup.picasso.Picasso
import com.squareup.picasso.Target
import timber.log.Timber

class ChannelsDropdownAdapter(private val channelList: List<Channel>, context: Context) : ArrayAdapter<Channel>(context, 0, channelList) {

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val c = channelList[position]
        var view = convertView
        val viewHolder: ViewHolder

        if (view == null) {
            val binding = StaxSpinnerItemWithLogoBinding.inflate(LayoutInflater.from(parent.context), parent, false)
            view = binding.root
            viewHolder = ViewHolder(binding)
        } else
            viewHolder = view.tag as ViewHolder

        viewHolder.setChannel(c)
        return view
    }

    override fun getCount(): Int = channelList.size

    override fun getItemId(position: Int): Long = position.toLong()

    override fun getItem(position: Int): Channel? = if (channelList.isEmpty()) null else channelList[position]

    inner class ViewHolder(val binding: StaxSpinnerItemWithLogoBinding) : Target {

        fun setChannel(channel: Channel) {
            binding.serviceItemNameId.text = channel.toString()
            UIHelper.loadPicasso(channel.logoUrl, Constants.size55, this)
        }

        override fun onBitmapLoaded(bitmap: Bitmap?, from: Picasso.LoadedFrom?) {
            val d = RoundedBitmapDrawableFactory.create(binding.root.context.resources, bitmap)
            d.isCircular = true
            binding.serviceItemImageId.setImageDrawable(d)
        }

        override fun onBitmapFailed(e: Exception?, errorDrawable: Drawable?) {
            Timber.e(e)
        }

        override fun onPrepareLoad(placeHolderDrawable: Drawable?) {}

    }
}